### ISSUE TITLE
Appimage hardware decoding doesn't work

### ISSUE THREAD

**Issue by Sunspark-007** [mpc-qt/mpc-qt#308]
Hi, version 24.12. This isn't libmpv's fault because hardware decoding works in other players (as well as Haruna which also uses libmpv). You can see that the CPU usage is higher despite the checkbox being checked.

All the other players I am using are flatpaks.. mpc-qt is the only appimage one, so I am not sure if this is playing a role or not.

I would also suggest introducing a key to allow people to toggle hw decoding on and off like how mpv uses ctrl-h to do so, and you can see the output of that reflected in the OSD.

Operating System: SteamOS 3.6.20 Build: 20241030.1
KDE Plasma Version: 5.27.10
KDE Frameworks Version: 5.115.0
Qt Version: 5.15.12
Kernel Version: 6.5.0-valve22-1-neptune-65-g9a338ed8a75e (64-bit)
Graphics Platform: X11

**Comment by tbertels** [comment]
Thank you for the report!

I can reproduce with the 24.12 and the 24.06 AppImage versions.

Based on the libmpv log, it's trying to access the hardware decoding library in
 `/usr/lib/x86_64-linux-gnu/dri/`
instead of
`/usr/lib/dri/`

So it seems as if the AppImage version changed `LIBVA_DRIVERS_PATH`.

**Comment by Sunspark-007** [comment]
I have discovered that Fedora's flatpak repository has this player (Flathub does not). Hardware decoding does work with the flatpak version from Fedora.

This is an option to use.

**Comment by tbertels** [comment]
> I have discovered that Fedora's flatpak repository has this player (Flathub does not). Hardware decoding does work with the flatpak version from Fedora.
> 
> This is an option to use.

That's great to hear!

Does x265 hardware/software decoding work too?
That's an issue we've been having in #31 along with eac3 decoding.

**Comment by Sunspark-007** [comment]
> Does x265 hardware/software decoding work too?

No. :( I thought it would, but it can't even do software decoding of x265. Doesn't open at all, even with the decoding box unchecked.

I assumed it was present because vaapi was showing up for vp9.

**Comment by Sunspark-007** [comment]
@tbertels Hi, trying v24.12.1 to see how it does now.

h.264 - vaapi-copy
h.265 - vaapi-copy
vp09 - software decoding
av1 - software decoding

**Comment by tbertels** [comment]
Indeed, the AppImage version of FFmpeg doesn't include the hardware versions of the codecs:

Here's with the normal version:
>[1.807] [vd] v: Codec list:
>[1.807] [vd] v:     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
>[1.807] [vd] v:     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
>[1.807] [vd] v:     h264_qsv (h264) - H264 video (Intel Quick Sync Video acceleration)
>[1.807] [vd] v:     h264_cuvid (h264) - Nvidia CUVID H264 decoder

And here's with the AppImage version:
>[2.330] [vd] v: Codec list:
>[2.330] [vd] v:     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
>[2.330] [vd] v:     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper

So both `h264_qsv` and `h264_cuvid` are missing.

For `h264_qsv`, [here's how to do it](https://gist.github.com/andrearug/ee461152acfbbd3f52b3bb9ffb55b6b8) (with Ubuntu 16.04).
So it's neither trivial nor up to date. And the bad news doesn't stop here:
Intel has abandoned MediaSDK and has switched to [oneAPI](https://www.intel.com/content/www/us/en/developer/articles/guide/get-started-with-the-oneapi-video-processing-library.html) for newer architectures.

There's a similar [procedure for](https://developer.nvidia.com/blog/nvidia-ffmpeg-transcoding-guide/) `h264_cuvid`.

I don't know about AMD.

So it's rather involved and not my priority right now. It would also increase the frailty of the AppImage maker CI.
I'd rather fix the Flatpak.
But if someone is willing to work on this, PRs are welcome.

**Comment by Sunspark-007** [comment]
It's not a priority, but I am a little mystified though because I don't have insight on the technical challenges involved.. why are you guys having trouble with the flatpak when all the other players have flatpaks with hardware decoding for all codecs?

**Comment by tbertels** [comment]
With the (now working) Flatpak, I've noticed that `h264_qsv (h264) - H264 video (Intel Quick Sync Video acceleration)` isn't actually needed for hardware decoding on Intel.
Here's the log from the Flatpak:

```
[67.696] [vd] v: Codec list:
[67.696] [vd] v:     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
[67.696] [vd] v:     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
[67.696] [vd] v:     h264_cuvid (h264) - Nvidia CUVID H264 decoder
[67.696] [vd] v: Opening decoder h264
[67.696] [vd] v: Looking at hwdec h264-nvdec...
[67.696] [vd] v: Could not create device.
[67.696] [vd] v: Looking at hwdec h264-vaapi...
[67.696] [vd] v: Trying hardware decoding via h264-vaapi.
[67.757] [vd] info: Using hardware decoding (vaapi).

```

So maybe there's a way to make it work with the AppImage as well.
